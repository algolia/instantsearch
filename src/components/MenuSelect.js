import React, { Component } from 'preact-compat';
import PropTypes from 'prop-types';

import Template from './Template';
import autoHideContainerHOC from '../decorators/autoHideContainer';
import headerFooterHOC from '../decorators/headerFooter';

class MenuSelect extends Component {
  static propTypes = {
    cssClasses: PropTypes.shape({
      select: PropTypes.string,
      option: PropTypes.string,
    }),
    items: PropTypes.array.isRequired,
    refine: PropTypes.func.isRequired,
    templateProps: PropTypes.object.isRequired,
  };

  handleSelectChange = ({ target: { value } }) => {
    this.props.refine(value);
  };

  render() {
    const { cssClasses, templateProps, items } = this.props;
    const { value: selectedValue } = items.find(item => item.isRefined) || {
      value: '',
    };

    return (
      <select
        className={cssClasses.select}
        value={selectedValue}
        onChange={this.handleSelectChange}
      >
        <option value="" className={cssClasses.option}>
          <Template templateKey="seeAllOption" {...templateProps} />
        </option>

        {items.map(item => (
          <option
            key={item.value}
            value={item.value}
            className={cssClasses.option}
          >
            <Template data={item} templateKey="item" {...templateProps} />
          </option>
        ))}
      </select>
    );
  }
}

export default autoHideContainerHOC(headerFooterHOC(MenuSelect));
