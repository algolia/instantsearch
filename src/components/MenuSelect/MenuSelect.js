import React, { Component } from 'preact-compat';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Template from '../Template/Template';

class MenuSelect extends Component {
  static propTypes = {
    cssClasses: PropTypes.shape({
      root: PropTypes.string.isRequired,
      select: PropTypes.string.isRequired,
      option: PropTypes.string.isRequired,
    }).isRequired,
    items: PropTypes.array.isRequired,
    refine: PropTypes.func.isRequired,
    templateProps: PropTypes.object.isRequired,
  };

  handleSelectChange = ({ target: { value } }) => {
    this.props.refine(value);
  };

  render() {
    const { cssClasses, templateProps, items } = this.props;
    const { value: selectedValue } = items.find(item => item.isRefined) || {
      value: '',
    };

    const rootClassNames = cx(cssClasses.root, {
      [cssClasses.noRefinementRoot]: items.length === 0,
    });

    return (
      <div className={rootClassNames}>
        <select
          className={cssClasses.select}
          value={selectedValue}
          onChange={this.handleSelectChange}
        >
          <Template
            {...templateProps}
            templateKey="defaultOption"
            rootTagName="option"
            rootProps={{
              value: '',
              className: cssClasses.option,
            }}
          />

          {items.map(item => (
            <Template
              {...templateProps}
              templateKey="item"
              rootTagName="option"
              rootProps={{
                value: item.value,
                className: cssClasses.option,
              }}
              key={item.value}
              data={item}
            />
          ))}
        </select>
      </div>
    );
  }
}

export default MenuSelect;
