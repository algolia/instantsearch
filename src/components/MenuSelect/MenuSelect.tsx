/** @jsx h */

import { h } from 'preact';
import cx from 'classnames';
import { find } from '../../lib/utils';
import Template from '../Template/Template';

type MenuSelectTemplates = {
  defaultOption: string;
  item: string;
};

type MenuItem = {
  /**
   * The value of the menu item.
   **/
  value: string;
  /**
   * Human-readable value of the menu item.
   **/
  label: string;
  /**
   * Number of results matched after refinement is applied.
   **/
  count: number;
  /**
   * Indicates if the refinement is applied.
   **/
  isRefined: boolean;
};

type Props = {
  cssClasses: {
    root: string;
    noRefinementRoot: string;
    select: string;
    option: string;
  };
  items: MenuItem[];
  refine: (value: MenuItem['value']) => void;
  templateProps: {
    templates: MenuSelectTemplates;
  };
};

function MenuSelect({ cssClasses, templateProps, items, refine }: Props) {
  const { value: selectedValue } = find(items, item => item.isRefined) || {
    value: '',
  };

  return (
    <div
      className={cx(cssClasses.root, {
        [cssClasses.noRefinementRoot]: items.length === 0,
      })}
    >
      <select
        className={cssClasses.select}
        value={selectedValue}
        onChange={event => {
          refine((event.target as HTMLSelectElement).value);
        }}
      >
        <Template
          {...templateProps}
          templateKey="defaultOption"
          rootTagName="option"
          rootProps={{
            value: '',
            className: cssClasses.option,
          }}
        />

        {items.map(item => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="option"
            rootProps={{
              value: item.value,
              className: cssClasses.option,
            }}
            key={item.value}
            data={item}
          />
        ))}
      </select>
    </div>
  );
}

export default MenuSelect;
