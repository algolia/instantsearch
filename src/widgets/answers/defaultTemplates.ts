import { AnswersComponentTemplates } from '../../components/Answers/Answers';

const defaultTemplates: AnswersComponentTemplates = {
  header: '',
  loader: '',
  item: (item) => JSON.stringify(item),
};

export default defaultTemplates;
