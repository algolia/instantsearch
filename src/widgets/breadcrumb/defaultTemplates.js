export default {
  home: 'Home',
  separator: ' > ',
};
