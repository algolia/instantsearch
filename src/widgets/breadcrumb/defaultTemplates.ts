import type { BreadcrumbComponentTemplates } from '../../components/Breadcrumb/Breadcrumb';

const defaultTemplates: BreadcrumbComponentTemplates = {
  home: 'Home',
  separator: '>',
};

export default defaultTemplates;
