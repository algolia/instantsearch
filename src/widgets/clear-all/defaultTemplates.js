export default {
  header: '',
  link: 'Clear all',
  footer: ''
};
