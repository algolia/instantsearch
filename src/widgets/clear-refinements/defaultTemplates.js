export default {
  button: 'Clear all',
};
