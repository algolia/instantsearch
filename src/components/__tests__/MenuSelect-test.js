import React from 'react';
import MenuSelect from '../MenuSelect';
import renderer from 'react-test-renderer';

import defaultTemplates from '../../widgets/menu-select/defaultTemplates';

describe('MenuSelect', () => {
  it('should render <MenuSelect /> with items', () => {
    const props = {
      items: [{ value: 'foo', label: 'foo' }, { value: 'bar', label: 'bar' }],
      refine: () => {},
      templateProps: { templates: defaultTemplates },
      shouldAutoHideContainer: false,
      cssClasses: {
        root: 'root',
        noRefinementRoot: 'noRefinementRoot',
        select: 'select',
        option: 'option',
      },
    };
    const tree = renderer.create(<MenuSelect {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render <MenuSelect /> with no items', () => {
    const props = {
      items: [],
      refine: () => {},
      templateProps: { templates: defaultTemplates },
      shouldAutoHideContainer: false,
      cssClasses: {
        root: 'root',
        noRefinementRoot: 'noRefinementRoot',
        select: 'select',
        option: 'option',
      },
    };
    const tree = renderer.create(<MenuSelect {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
