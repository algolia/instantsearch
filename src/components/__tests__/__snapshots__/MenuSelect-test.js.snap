// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuSelect should render <MenuSelect /> with items 1`] = `
<div
  style={
    Object {
      "display": "",
    }
  }
>
  <div
    className="ais-root"
  >
    <div
      className="ais-body"
    >
      <select
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "See all",
              }
            }
          />
        </option>
        <option
          value="foo"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "foo ()",
              }
            }
          />
        </option>
        <option
          value="bar"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "bar ()",
              }
            }
          />
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`MenuSelect should render with custom css classes 1`] = `
<div
  style={
    Object {
      "display": "",
    }
  }
>
  <div
    className="ais-root"
  >
    <div
      className="ais-body"
    >
      <select
        className="foo"
        onChange={[Function]}
        value=""
      >
        <option
          className="bar"
          value=""
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "See all",
              }
            }
          />
        </option>
        <option
          className="bar"
          value="foo"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "foo ()",
              }
            }
          />
        </option>
        <option
          className="bar"
          value="bar"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "bar ()",
              }
            }
          />
        </option>
      </select>
    </div>
  </div>
</div>
`;
